### PR TITLE
Allow side-effects to permissions

### PR DESCRIPTION
**Pull request by GuillaumeRx**
Closes: #1057 

This allows permissions to have side-effects. The side-effects are defined in the permission constraint and is handled after the user approved the permissions.

It allows permissions to have a `onPermitted` side-effect and a `onFailure` side-effect.

### CODE REVIEW

File: packages/permission-controller/src/PermissionController.ts
Comment (FrederikBolding) [inline packages/permission-controller/src/PermissionController.ts:1947]: I hadn't thought of the fact that multiple of the granted permissions could have side-effects. Perhaps we should disallow that?

Ideally we would only have one side-effect for the call, because otherwise how do we revert the effects of the first side-effect if the second fails?

File: packages/permission-controller/src/Permission.ts
Comment (Gudahtt) [inline packages/permission-controller/src/Permission.ts:356]: If these are to be async, we'll want to return `Promise<void>` instead I support

Comment (Gudahtt) [inline packages/permission-controller/src/Permission.ts:356]: Curious why these first two are mandatory. Do you think we'll always need them both for any permissions that have a side-effect? I can't think of any counter-examples.

Having a JSDoc block for these would be fantastic, so that we can offer advice on what these are intended to be used for, and we can make recommendations about cleanup and so on.

**Comment by GuillaumeRx** [comment]
@metamaskbot publish-preview

**Comment by GuillaumeRx** [comment]
@metamaskbot publish-preview

**Comment by GuillaumeRx** [comment]
@metamaskbot publish-preview

File: packages/permission-controller/src/PermissionController.ts
Comment (matthewwalsh0) [inline packages/permission-controller/src/PermissionController.ts:2039]: Forgive me if I'm missing the obvious, but could this not just be implemented directly in the `PermissionController` by just invoking the `onPermitted` handlers after a successful `messagingSystem.call` and then calling the `onFailure` handlers for any of those callbacks that fail?

I'm not sure what we're gaining by adding the side effect support to the `ApprovalController` as it already returns a promise so does it not already support the callbacks we or anyone else would need via `await` or `then` and `catch`?

My worry would be we're complicating its API as the user would have to choose between Promises or callbacks to control the flow.

File: packages/approval-controller/src/ApprovalController.ts
Comment (matthewwalsh0) [inline packages/approval-controller/src/ApprovalController.ts:33]: Could the `failureHandlers` be optional so you don't even need to provide the empty array?

File: packages/approval-controller/src/ApprovalController.ts
Comment (matthewwalsh0) [inline packages/approval-controller/src/ApprovalController.ts:283]: This doc is also applicable to `add` as it also calls `_add`.

File: packages/approval-controller/src/ApprovalController.ts
Comment (matthewwalsh0) [inline packages/approval-controller/src/ApprovalController.ts:32]: Is it overkill to have a `SideEffect` type for `(params: SideEffectParams) => Promise<unknown>` to avoid this duplication?

File: packages/permission-controller/src/PermissionController.ts
Comment (GuillaumeRx) [reply packages/permission-controller/src/PermissionController.ts:2039]: @Gudahtt and I discussed it a long time ago, IIRC it's to also support side-effects for non-permissions approvals (like adding a new chain, that was the example @Gudahtt gave me I think). It also implements the support in the PermissionController because we need it at this level for Snaps

File: packages/approval-controller/src/ApprovalController.ts
Comment (GuillaumeRx) [reply packages/approval-controller/src/ApprovalController.ts:33]: Initially we wanted to require both to be defined but I'm considering making it optional since failure cases can be handled before.

File: packages/approval-controller/src/ApprovalController.ts
Comment (GuillaumeRx) [reply packages/approval-controller/src/ApprovalController.ts:32]: Indeed it could be nice

File: packages/approval-controller/src/ApprovalController.ts
Comment (cryptotavares) [reply packages/approval-controller/src/ApprovalController.ts:467]: I would prefer an early retur, instead of an if/else return to be honest.

File: packages/permission-controller/src/PermissionController.ts
Comment (GuillaumeRx) [reply packages/permission-controller/src/PermissionController.ts:2049]: > But my main concern is I fear we're adding complexity to the API of the ApprovalController that it doesn't require to support the use cases we may want in future

I'll let @Gudahtt answer on this one

> where we could use native features like Promise.all with multiple approval requests for example
I don't see where this would resolve the issue side-effects need to resolve

Comment (FrederikBolding) [reply packages/permission-controller/src/PermissionController.ts:2049]: If the confirmations systems team would prefer just handling the side-effects in the PermissionController for now I would say that is fine. That still solves our problems and seems cleaner. 

Side-effects to approvals may be something that should be handled differently? Regardless, they will own adding that when it is needed, so 🤷 Would be nice with a unified system though..

File: packages/approval-controller/src/ApprovalController.ts
Comment (FrederikBolding) [reply packages/approval-controller/src/ApprovalController.ts]: Agreed

File: tsconfig.packages.json
Comment (FrederikBolding) [inline tsconfig.packages.json:5]: I'm not sure we can go beyond `ES2020` yet? 🤔

File: packages/approval-controller/src/ApprovalController.ts
Comment (FrederikBolding) [inline packages/approval-controller/src/ApprovalController.ts:477]: Why are the returned values different here? That seems confusing.

File: packages/permission-controller/src/PermissionController.ts
Comment (FrederikBolding) [inline packages/permission-controller/src/PermissionController.ts:2037]: Could we simplify by having the empty array for both in the accumulator? That way we can eliminate some ifs.

File: packages/approval-controller/src/ApprovalController.ts
Comment (FrederikBolding) [inline packages/approval-controller/src/ApprovalController.ts:720]:
```suggestion
    if (rejectedHandlers.length > 0) {
```

File: packages/approval-controller/src/ApprovalController.ts
Comment (FrederikBolding) [inline packages/approval-controller/src/ApprovalController.ts:741]:
```suggestion
            'Multiple errors occurred during side-effects execution',
```

**Review by Mrtenz — CHANGES_REQUESTED** [review]
Please add a PR description rather than just linking to an issue.

File: packages/permission-controller/src/PermissionController.ts
Comment (GuillaumeRx) [reply packages/permission-controller/src/PermissionController.ts:2049]: As per today's meeting we will handle side-effects in the PermissionController for now.

File: packages/approval-controller/src/ApprovalController.ts
Comment (GuillaumeRx) [reply packages/approval-controller/src/ApprovalController.ts]: This will be not relevant anymore with the handling of side-effects on the PermissionController side

File: packages/permission-controller/src/Permission.ts
Comment (FrederikBolding) [inline packages/permission-controller/src/Permission.ts:382]:
```suggestion
   * A method triggered if a `onPermitted` method rejected.
```

File: packages/permission-controller/src/PermissionController.ts
Comment (FrederikBolding) [inline packages/permission-controller/src/PermissionController.ts:2082]: Does this need to be async?

File: packages/permission-controller/src/PermissionController.test.ts
Comment (FrederikBolding) [inline packages/permission-controller/src/PermissionController.test.ts:3197]:
```suggestion
        Promise.reject(new Error('error')),
```

File: packages/permission-controller/src/PermissionController.test.ts
Comment (FrederikBolding) [inline packages/permission-controller/src/PermissionController.test.ts:3252]:
```suggestion
        Promise.reject(new Error('error')),
```

File: packages/permission-controller/src/PermissionController.test.ts
Comment (FrederikBolding) [inline packages/permission-controller/src/PermissionController.test.ts:3300]:
```suggestion
        Promise.reject(new Error('error')),
```

File: packages/permission-controller/src/PermissionController.test.ts
Comment (FrederikBolding) [inline packages/permission-controller/src/PermissionController.test.ts:3304]:
```suggestion
        Promise.reject(new Error('error')),
```

File: packages/permission-controller/src/PermissionController.ts
Comment (GuillaumeRx) [reply packages/permission-controller/src/PermissionController.ts:2094]: We cannot merge the two if statements since it would mean adding an `else if` statemenent for all the things done ing there is a rejectedHandler but no failureHandlers

Comment (GuillaumeRx) [reply packages/permission-controller/src/PermissionController.ts:2094]: I moved the try catch block under the second if statement though

**Comment by GuillaumeRx** [comment]
@metamaskbot publish-preview

File: packages/permission-controller/src/PermissionController.ts
Comment (FrederikBolding) [reply packages/permission-controller/src/PermissionController.ts]: Ah, yeah. This is hard to read 😅

**Comment by GuillaumeRx** [comment]
@metamaskbot publish-preview

File: packages/permission-controller/src/PermissionController.ts
Comment (matthewwalsh0) [inline packages/permission-controller/src/PermissionController.ts:2102]: Currently it seems like we're running all the failure handers if any of the permitted handlers fail, so just one rejection would run all the failure handlers, even for the successful permitted handlers. Is this intentional?

File: packages/permission-controller/src/PermissionController.ts
Comment (matthewwalsh0) [inline packages/permission-controller/src/PermissionController.ts:1923]: It seems like the `executeSideEffects` function returns an array of side effect results, for all the provided onPermitted handlers. But here, it looks like we are mapping those results against each permission by index, even if that permission didn't have any side effects.

So as an example, if you requested 10 permissions, and only the last 5 had side effects, would the first five without side effects be assigned the side effect data, and the remaining 5 with side effects get undefined?

File: packages/permission-controller/src/PermissionController.ts
Comment (GuillaumeRx) [reply packages/permission-controller/src/PermissionController.ts:2102]: Yes this is intentional. Requested permissions can be inter-dependent, same for side-effects. Since permissions are not granted if a side-effect fails `onFailure` will have to rollback any changes made for all the side-effects

File: packages/permission-controller/src/PermissionController.ts
Comment (GuillaumeRx) [reply packages/permission-controller/src/PermissionController.ts:1923]: Good catch !

**Review by matthewwalsh0 — APPROVED** [review]

**Review by FrederikBolding — APPROVED** [review]
LGTM - please make sure it works locally together with the snaps changes before merging

**Comment by GuillaumeRx** [comment]
@metamaskbot publish-preview

**Pull request merged**
